crates/subspace-runtime-primitives/src/lib.rs

```diff
@@ -48,15 +74,66 @@ pub type Moment = u64;
 /// of data like extrinsics, allowing for them to continue syncing the network through upgrades
 /// to even the core data structures.
 pub mod opaque {
-    use super::BlockNumber;
-    use sp_runtime::generic;
-    use sp_runtime::traits::BlakeTwo256;
-    pub use sp_runtime::OpaqueExtrinsic as UncheckedExtrinsic;
+    use super::{BlockNumber, RECORDED_HISTORY_SEGMENT_SIZE};
+    use parity_scale_codec::{Decode, Encode};
+    #[cfg(feature = "std")]
+    use serde::{Deserialize, Serialize};
+    use sp_core::RuntimeDebug;
+    use sp_runtime::traits::{BlakeTwo256, Block as BlockT, Header as HeaderT};
+    use sp_runtime::{generic, DigestItem, OpaqueExtrinsic};
+    use sp_std::prelude::*;
 
     /// Opaque block header type.
     pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
     /// Opaque block type.
-    pub type Block = generic::Block<Header, UncheckedExtrinsic>;
+
+    /// Abstraction over a substrate block.
+    #[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug)]
+    #[cfg_attr(
+        feature = "std",
+        derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf)
+    )]
+    #[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
+    #[cfg_attr(feature = "std", serde(deny_unknown_fields))]
+    pub struct Block {
+        /// The block header.
+        pub header: Header,
+        /// The accompanying extrinsics.
+        pub extrinsics: Vec<OpaqueExtrinsic>,
+    }
+
+    impl BlockT for Block {
+        type Extrinsic = OpaqueExtrinsic;
+        type Header = Header;
+        type Hash = <Header as HeaderT>::Hash;
+
+        fn header(&self) -> &Self::Header {
+            &self.header
+        }
+        fn extrinsics(&self) -> &[Self::Extrinsic] {
+            &self.extrinsics[..]
+        }
+        fn deconstruct(self) -> (Self::Header, Vec<Self::Extrinsic>) {
+            (self.header, self.extrinsics)
+        }
+        fn new(mut header: Self::Header, extrinsics: Vec<Self::Extrinsic>) -> Self {
+            if header.number == 0 {
+                // This check is necessary in case block was deconstructed and constructed again.
+                if header.digest.logs.is_empty() {
+                    // We fill genesis block with extra data such that the very first archived
+                    // segment can be produced right away, bootstrapping the farming process.
+                    let ballast = vec![0; RECORDED_HISTORY_SEGMENT_SIZE as usize];
+                    header.digest.logs.push(DigestItem::Other(ballast));
```

Comment:
If this abstraction is merely for injecting a `DigesteItem`, why you can't simply call `deposit_log` in genesis_build? 

By filling the block itself, what pieces do we need to fill in? If it's all about the genesis state and the digests in the header, I'm pretty sure you could achieve that in the genesis building process.

Reply:
`GenesisBuild` is just for storage such that genesis block can include its root. There is no block production as such and no hooks are running in pallets on genesis block: https://github.com/paritytech/substrate/blob/1d818f38a3198f21b4ae68bac3dc98219fccc669/client/service/src/client/client.rs#L338-L342

Reply:
Thanks for the link, I had some misunderstandings here. I think it's a reasonable feature request for Substrate, I'll file one there later.